### ISSUE TITLE
An eonBuffer crash during Chapter2 "Lost Lambs" while going through the canyon.

### ISSUE THREAD

**Issue by psi29a** [KillaW0lf04/The-Witcher-2-Issues#82]
This is reproducible, every time I go through the canyon there in Chapter2, it crashes to desktop.

This also happens when playing around (looking at the water) in front of the fort next to the small bridge.

<pre>
witcher2: /mnt/Storage/Development/Projects/eon2-witcher2/src/dlls/directx/d3d9/eonBuffer.cpp:445: DWORD eON_Buffer::OpenGLSide_Unlock(DWORD, LPVOID): Assertion `mBufferPtr' failed.
</pre>

**Comment by vpltd-peterm** [comment]
Does this still occur with the latest beta?

**Comment by psi29a** [comment]
Hello again, and thank you for asking. I tried it out immediatly. Around the same spot as the original crash, the game hit the disk pretty hard. It took it a few minutes before becoming responsive again... which is an improvement over crashing. However moving deeper into the cliffs/canyon area it finally did crash, without a handler, with this:

<pre>
/home/bcurtis/.steam/steam/SteamApps/common/the witcher 2/CrashReporter: error while loading shared libraries: libgtk-3.so.0: cannot open shared object file: No such file or directory
Installing breakpad exception handler for appid(steam)/version(1422054110)
Game removed: AppID 20920 "The Witcher 2: Assassins of Kings Enhanced Edition", ProcID 24793 
</pre>

The file it is looking for exists:
<pre>
bcurtis@Wintermute:~$ find /usr/lib | grep libgtk-3.so.0
/usr/lib/x86_64-linux-gnu/libgtk-3.so.0
/usr/lib/x86_64-linux-gnu/libgtk-3.so.0.1000.8
</pre>

I'm assuming that it isn't looking in the usual places? Maybe it needs the 32-bit libgtk-3? I can try again tomorrow.

Do you know where the crash logs are kept or even if it is dumped to disk if the crash catcher doesn't execute?

**Comment by vpltd-peterm** [comment]
Yes, it will require the 32 bit version of GTK3. Damn... I do have a 64 bit build of the crash reporter, so I guess it's time to look at making some sort of "universal binary".

The game's log file is ~/.local/share/cdprojektred/witcher2/eon.txt - there may also be a .dmp file here which is a Google Breakpad minidump. Unlike the previous system, these are now sent directly to us, rather than a text-only stacktrace.

**Comment by psi29a** [comment]
OK, I have the libgtk-3-0:i386 installed... but it seems that witcher2 ever reaching this is limited. Most of the time, it pages out to swap for 2~5 minutes then the OOM killer comes along and nukes the process.

<pre>
[296185.922003] Out of memory: Kill process 27994 (E.runloop) score 362 or sacrifice child
[296185.922005] Killed process 27994 (E.runloop) total-vm:3758088kB, anon-rss:3017804kB, file-rss:250028kB
[296185.922319] Purging GPU memory, 0 bytes freed, 6496256 bytes still pinned.
</pre>

This is a system with 8GiB of ram and nothing going on aside steam. No dmp file.
<pre>
drwx------ 3 bcurtis bcurtis   4096 Jan 25 08:56 .
drwx------ 3 bcurtis bcurtis   4096 Aug 12 12:07 ..
-rw-rw-r-- 1 bcurtis bcurtis 706012 Jan 25 08:58 eONprecompiledShaders32.dat
-rw-rw-r-- 1 bcurtis bcurtis  22507 Jan 25 08:56 EonRegistry.xml
-rw-rw-r-- 1 bcurtis bcurtis   4218 Jan 25 08:57 eon.txt
drwx------ 3 bcurtis bcurtis   4096 Aug 12 12:07 GameDocuments
</pre>

Would you like a copy of the save-game I'm using?  Here is the eon.txt:
https://gist.githubusercontent.com/psi29a/6f572a5458e9bc7c8ab2/raw/e7e2c27b10f38b1c7468b55c239552949220305d/gistfile1.txt

I'll keep trying to get a crash report... 

Anything else I can try?

**Comment by psi29a** [comment]
Oh fun.... so the crash reporter can't run because of something else:
<pre>
wwwwERROR: ld.so: object '/home/bcurtis/.steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
/home/bcurtis/.steam/steam/SteamApps/common/the witcher 2/CrashReporter: /home/bcurtis/.steam/ubuntu12_32/steam-runtime/i386/usr/lib/i386-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /home/bcurtis/.steam/steam/SteamApps/common/the witcher 2/CrashReporter)
wwwInstalling breakpad exception handler for appid(steam)/version(1422054110)
Game removed: AppID 20920 "The Witcher 2: Assassins of Kings Enhanced Edition", ProcID 28557 
</pre>

Seriously?
<pre>
bcurtis@Wintermute:~$ strings /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBC
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBC_2.3
GLIBC_2.2.5
GLIBC_2.14
GLIBC_2.4
GLIBC_2.3.4
GLIBC_2.17
GLIBC_2.3.2
GLIBCXX_DEBUG_MESSAGE_LENGTH
</pre>

This system is Ubuntu Trusty LTS, which is supported until April 2019.  

No plans to update to Utopic because it is a crap release with lots of broken parts.

**Comment by vpltd-peterm** [comment]
OK will look into linkng to GLIBCXX_3.4 rather than 3.4.20 explicitly
The saved game would be useful, to see if i can locally reproduce this problem

**Comment by psi29a** [comment]
Sounds great! :)

Here is the save game:
https://dl.dropboxusercontent.com/u/396161/Lost_Lambs.tar.bz2

**Comment by psi29a** [comment]
With the save, just move forward into the path a bit and hang a left... It
should then show the loading snake. At this point it either stops and
eventually is
killed by the OOM Killer or pauses a bit then you can move along. Going to
right in the direction of the beach will crash it to desktop. At least for
me. Would be nice if others here can also test the save. :)

**Comment by licaon-kter** [comment]
Can't repro these with your savegame on my Debian Sid / 64bit / kernel 3.18 / nVidia 346.35.
Went front left small load snake no issue, load savegame again went right -> beach -> huge fps drop from 50 to 10 on snake load but after that it recovered back. No crash, no oom.

**Comment by psi29a** [comment]
3.16.0-29-generic here, but I'll also give a newer version of drivers a go when I get a chance. To be fair, I've updated a few times now and it hasn't advanced the plot one bit. ;)

**Comment by psi29a** [comment]
I saw this (opengl21) in another thread and also gave it a shot:
<pre>
bcurtis@Wintermute:~/.steam/steam/SteamApps/common/the witcher 2$ LD_LIBRARY_PATH=~/.steam/bin32/ ~/.steam/bin32/steam-runtime/run.sh ./witcher2 --eon_opengl21
Setting breakpad minidump AppID = 20920
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561197971193685 [API loaded no]
/home/bcurtis/.steam/steam/SteamApps/common/the witcher 2/CrashReporter: /home/bcurtis/.steam/bin32/steam-runtime/i386/usr/lib/i386-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /home/bcurtis/.steam/steam/SteamApps/common/the witcher 2/CrashReporter)
</pre>

Looks like fun... :)

*waits patiently*

**Comment by vpltd-peterm** [comment]
That's what I get for compiling the crash reporter with GCC 4.9 ;) With GCC 4.8 there should be no problem.

**Comment by DerRidda** [comment]
Not a fix but a workaround. I saw the same issues and moving old saves out of the saves folder helped remedy this. This might in the end be related to the stability issues The Witcher 2 has with lots of saves on Windows as well.

On Windows there is some registry-fu you can do to help with that and I do not know if VP could reproduce that with eON.

**Comment by psi29a** [comment]
Thanks @DerRidda, I'll give it a try and see what happens.

**Comment by vpltd-peterm** [comment]
WE could, depending what the registry settings do. Got a link ? :)

It almost sounds as if the game is trying to allocate a stupidly large vertex buffer or so, eON tries to do it and we get OOM'd as a result. We'd like to look into it and see if it's a bug with Witcher 2 or a bug with us.

So far, I cannot reproduce the behaviour from your saved game though.

**Comment by psi29a** [comment]
I'm using steam cloud, which makes this hard to test a bit... but I moved 700MiB of save games out of the way except for the last 5 (combination autosave and manualsave).  This seems to have worked as I no longer have game crashing issues.

I'm thinking that perhaps the original bug was either fixed by the latest release OR triggered by having too many save files.  The dumps in the beginning of this bug report seem to be in no way related to saved-games. It wasn't until tested the new releases that the behaviour has changed.

I would still keep this bug open until the crash catcher/reporter is fixed. I can reliable test it by using the opengl21 flag. ;)

**Comment by psi29a** [comment]
Crazy, I know... but would anyone like a 700MiB tarball (bz2) to test then? ;)  I might be able to throw it up on DropBox without them complaining too much.

**Comment by DerRidda** [comment]
@vpltd-peterm Seems like I remembered it wrong: http://pcgamingwiki.com/wiki/The_Witcher_2:_Assassins_of_Kings#Black_Screen_During_the_Cut-scene_after_the_First_Chapter_.28Enhanced_Edition.29

It seem there's nothing you can do to fix it really, just remedy it with that one setting and by periodically deleting/moving saves.

**Comment by vpltd-peterm** [comment]
Ah,  i see the  increaseUserVa 3072 hint to give 3GB address space...
This doesn't really apply to eON.. we don't emulate the hard 2GB address space limit. So if Linux is capable of allocating 3GB+ in a 32 bit process, so are we.

We agree, the leak is annoying, but unfortunately there is nothing we can do our side to fix it..

**Comment by psi29a** [comment]
Just tested latest build.

I can't seem to get the crash reporter to trigger now... which is a good thing I guess. ;)

Do you know of a way to trigger it to test that the GLIBC issue is taken care of?

**Comment by vpltd-peterm** [comment]
Run CrashReporter binary from the shell with "test.appid" as the only parameter should do.. or check it with ldd :)

**Comment by psi29a** [comment]
<pre>
bcurtis@Wintermute:~/.steam/steam/SteamApps/common/the witcher 2$ ./CrashReporter test.appid
Handling crash report for AppID: test.appid
Loaded libcurl: libcurl.so.4
libcurl version: libcurl/7.37.1 OpenSSL/1.0.1f zlib/1.2.8 libidn/1.28 librtmp/2.3
</pre>

Wrapper script around the i386 and x86_64 binaries, works for me. :)

Cheers!

For the late comers here:
TLDR; Delete or move most of your saves and the crashes go away. This is a bug in Witcher2, not in eON.
